src/adapters/mqtt/index.ts

```diff
@@ -1,124 +1,175 @@
+            serverBinding.lastWill.qos
+            ? serverBinding.lastWill.qos
+            : undefined,
+        payload:
+          serverBinding &&
+            serverBinding.lastWill &&
+            serverBinding.lastWill.message
+            ? serverBinding.lastWill.message
+            : undefined,
+        retain:
+          serverBinding &&
+            serverBinding.lastWill &&
+            serverBinding.lastWill.retain
+            ? serverBinding.lastWill.retain
+            : undefined,
+      },
+      keepalive: serverBinding && serverBinding.keepAlive,
+      username: userAndPasswordSecurityReq
+        ? mqttOptions?.authentication?.userPassword.username
+        : undefined,
+      password: userAndPasswordSecurityReq
+        ? mqttOptions?.authentication?.userPassword.password
+        : undefined,
+      ca: X509SecurityReq ? mqttOptions?.authentication?.cert : undefined,
+    } as any)
+
+
+    this.client.on("message", (channel, message, mqttPacket) => {
+      const msg = this._createMessage(mqttPacket as IPublishPacket)
+      this.emit("message", msg, this.client)
+    })
 
-        resolve(this)
+    this.client.on("reconnect", () => {
+      this.emit("reconnect", {
+        connection: this.client,
+        channels: this.channelNames,
       })
+    })
 
-      this.client.on('message', (channel, message, mqttPacket) => {
-        const msg = this._createMessage(mqttPacket as IPublishPacket)
-        this.emit('message', msg, this.client)
+    this.client.on("close", () => {
+      this.emit("close", {
+        connection: this.client,
+        channels: this.channelNames,
       })
+    })
 
-      this.client.on('reconnect', () => {
-        this.emit('reconnect', {
-          connection: this.client,
-          channels: this.channelNames,
-        })
-      })
-      
-      this.client.on('close', () => {
-        this.emit('close', {
-          connection: this.client,
-          channels: this.channelNames,
+    this.client.on("error", (error) => {
+      this.emit("error", error)
+    })
+
+
+    const connectClient = () => {
+      return new Promise((resolve) => {
+        this.client.on("connect", () => {
+          if (!this.firstConnect) {
+            this.firstConnect = true
+            this.emit("connect", {
+              name: this.name(),
+              adapter: this,
+              connection: this.client,
+              channels: this.channelNames,
+            })
+          }
+
+          if (Array.isArray(subscribedChannels)) {
+            subscribedChannels.forEach((channel) => {
+              const operation = this.parsedAsyncAPI.channel(channel).publish()
+              const binding = operation.binding("mqtt")
+              this.client.subscribe(channel, {
+                qos: binding && binding.qos ? binding.qos : 0,
+              })
+            })
+          }
+
+          resolve(this)
         })
       })
+    }
+
+    return connectClient()
```

Comment:
I am trying to promisify this function, but getting this error -> 
```
Type 'unknown' is not assignable to type 'this'.
  'this' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.
```

I am most probably missing something. @fmvilas can you help me on this.

Reply:
Solved it, was missing the return type.